kaldb/src/main/java/com/slack/kaldb/bulkIngestApi/BulkIngestKafkaProducer.java

```diff
@@ -0,0 +1,298 @@
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.errors.AuthorizationException;
+import org.apache.kafka.common.errors.OutOfOrderSequenceException;
+import org.apache.kafka.common.errors.ProducerFencedException;
+import org.apache.kafka.common.errors.TimeoutException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BulkIngestKafkaProducer extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(BulkIngestKafkaProducer.class);
+
+  private final KafkaProducer<String, byte[]> kafkaProducer;
+
+  private final KafkaClientMetrics kafkaMetrics;
+
+  private final KaldbConfigs.KafkaConfig kafkaConfig;
+
+  private final DatasetMetadataStore datasetMetadataStore;
+  private final KaldbMetadataStoreChangeListener<DatasetMetadata> datasetListener =
+      (_) -> cacheSortedDataset();
+
+  protected List<DatasetMetadata> throughputSortedDatasets;
+
+  private final BlockingQueue<BulkIngestRequest> pendingRequests;
+
+  private final Integer producerSleep;
+
+  public static final String FAILED_SET_RESPONSE_COUNTER =
+      "bulk_ingest_producer_failed_set_response";
+  private final Counter failedSetResponseCounter;
+  public static final String STALL_COUNTER = "bulk_ingest_producer_stall_counter";
+  private final Counter stallCounter;
+
+  public static final String BATCH_SIZE_GAUGE = "bulk_ingest_producer_batch_size";
+  private final AtomicInteger batchSizeGauge;
+
+  private static final Set<String> OVERRIDABLE_CONFIGS =
+      Set.of(
+          ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
+
+  public BulkIngestKafkaProducer(
+      final DatasetMetadataStore datasetMetadataStore,
+      final KaldbConfigs.PreprocessorConfig preprocessorConfig,
+      final PrometheusMeterRegistry meterRegistry) {
+
+    this.kafkaConfig = preprocessorConfig.getKafkaConfig();
+
+    checkArgument(
+        !kafkaConfig.getKafkaBootStrapServers().isEmpty(),
+        "Kafka bootstrapServers must be provided");
+
+    checkArgument(!kafkaConfig.getKafkaTopic().isEmpty(), "Kafka topic must be provided");
+
+    this.datasetMetadataStore = datasetMetadataStore;
+    this.pendingRequests = new LinkedBlockingQueue<>();
+
+    // todo - consider making this a configurable value or removing the config
+    this.producerSleep =
+        Integer.parseInt(System.getProperty("kalDb.bulkIngest.producerSleep", "50"));
```

Comment:
perhaps we can add a unit suffix?

Reply:
I ended up adding a suffix to the variable name so it matches our current configs approach. I think long term we would benefit with a duration implementation that allows us to use arbitrary units in our configs (ie `50ms`), but a problem for another day.